### PR TITLE
[SOL-1636] Refactor get_voucher_from_code

### PR DESCRIPTION
**Pull request by vkaracic**
When no voucher exists for given code and when no products are associated with the voucher, raise exceptions instead of returning None.

This is a fix for: https://openedx.atlassian.net/browse/SOL-1636

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=644&repo=edx%2Fecommerce

**Comment by vkaracic** [comment]
@mjfrey please review.

**Comment by vkaracic** [comment]
I've added try..except blocks to preserve the existing error messages.

File: ecommerce/coupons/views.py [hunk 1]
```diff
@@ -114,7 +114,13 @@ def get_context_data(self, **kwargs):
         footer = get_lms_footer()
         code = self.request.GET.get('code', None)
         if code is not None:
-            voucher, product = get_voucher_from_code(code=code)
+            try:
+                voucher, product = get_voucher_from_code(code=code)
+            except Voucher.DoesNotExist:
```
Comment (mjfrey) [inline ecommerce/coupons/views.py:119]: should you add exceptions.ProductNotFoundError to this line?

File: ecommerce/coupons/views.py [hunk 2]
```diff
@@ -182,8 +188,11 @@ def get(self, request):
 
         if not code:
             return render(request, template_name, {'error': _('Code not provided')})
+        try:
+            voucher, product = get_voucher_from_code(code=code)
+        except Voucher.DoesNotExist:
```
Comment (mjfrey) [inline ecommerce/coupons/views.py:198]: same here.

**Comment by vkaracic** [comment]
@mjfrey I've added an exception catch for `ProductNotFoundError` and I've set it to return `No courses associated with coupon` as the error message.

**Comment by mjfrey** [comment]
:+1:

**Comment by vkaracic** [comment]
The project test coverage was failing even though the patch coverage is at 100% so I've added a couple of tests that are unrelated to boost that coverage. I don't want a negative project test coverage to affect other PR's.

**Comment by mjfrey** [comment]
:+1: 
Please squash

**Comment by vkaracic** [comment]
I'd like to keep the commits separate because they aren't related (the two last ones are just tests for unrelated code to keep the project coverage).

File: ecommerce/coupons/views.py [hunk 3]
```diff
@@ -182,8 +193,13 @@ def get(self, request):
...
+            return render(request, template_name, {'error': _('Coupon does not exist')})
+        except exceptions.ProductNotFoundError:
+            return render(request, template_name, {'error': _('No courses associated with coupon')})
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: No products are associated with this coupon.

File: ecommerce/coupons/views.py [hunk 4]
```diff
@@ -182,8 +193,13 @@ def get(self, request):
 
         if not code:
             return render(request, template_name, {'error': _('Code not provided')})
+        try:
+            voucher, product = get_voucher_from_code(code=code)
+        except Voucher.DoesNotExist:
+            return render(request, template_name, {'error': _('Coupon does not exist')})
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: The coupon is not valid.

File: ecommerce/extensions/fulfillment/tests/test_api.py [hunk 5]
```diff
@@ -51,6 +52,14 @@ def test_fulfill_order_incorrect_module(self):
         self.assertEquals(ORDER.FULFILLMENT_ERROR, self.order.status)
         self.assertEquals(LINE.FULFILLMENT_CONFIGURATION_ERROR, self.order.lines.all()[0].status)
 
+    @override_settings(FULFILLMENT_MODULES=['ecommerce.extensions.fulfillment.tests.modules.InvalidModule', ])
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_api.py]: It is not clear what you're trying to test here. What is an "invalid" module?

File: ecommerce/extensions/test/factories.py [hunk 6]
```diff
@@ -54,7 +54,10 @@ def prepare_voucher(code='COUPONTEST', _range=None, start_datetime=None, end_dat
         product = ProductFactory(title='Test product')
         _range = RangeFactory(products=[product, ])
     else:
-        product = _range.all_products()[0]
+        if _range.num_products() > 0:
```
Comment (clintonb) [inline ecommerce/extensions/test/factories.py]: 1. Combine with the `else` above.
2. Why is this necessary?

File: ecommerce/extensions/fulfillment/tests/test_api.py (hunk 5 above)
Comment (vkaracic) [inline ecommerce/extensions/fulfillment/tests/test_api.py]: It's a empty class module:
https://github.com/edx/ecommerce/pull/644/files#diff-2bfb94efa2bc715ec47e21d16132c336R66
used to trigger an unexpected error that would in turn trigger an exception catch here:
https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/fulfillment/api.py#L68
That's the line not covered by a test.

File: ecommerce/extensions/test/factories.py (hunk 6 above)
Comment (vkaracic) [inline ecommerce/extensions/test/factories.py]: It's for the `prepare_voucher` function to accept a range without a product:
https://github.com/edx/ecommerce/pull/644/files#diff-549419197ceb5146d6c1f4c27c33fee2R188

File: ecommerce/extensions/fulfillment/tests/test_api.py (hunk 5 above)
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_api.py]: This would be a bit cleaner if you patched a method on an existing fulfillment module, even the dummy one, to raise an error.

File: ecommerce/extensions/test/factories.py (hunk 6 above)
Comment (clintonb) [inline ecommerce/extensions/test/factories.py]: Okay. Merge into an `elif`.

**Comment by clintonb** [comment]
:+1: once comments addressed.

File: ecommerce/coupons/views.py (hunk 4 above)
Comment (clintonb) [inline ecommerce/coupons/views.py]: Actually, use the current message: No voucher found with code '??'.

File: ecommerce/coupons/views.py (hunk 3 above)
Comment (clintonb) [inline ecommerce/coupons/views.py]: Correction: The voucher is not applicable to your current basket.

**Comment by vkaracic** [comment]
@clintonb I believe I have addressed all your comments, can you take another look?

**Comment by clintonb** [comment]
:+1:
